### ISSUE TITLE
Tracking issue for implemented methods on `JsTypedArray`

### ISSUE THREAD

**Issue by jedel1043** [boa-dev/boa#3252]
Implementation in https://github.com/boa-dev/boa/blob/main/boa_engine/src/object/builtins/jstypedarray.rs

- [x] at
- [x] get buffer
- [x] get byteLength
- [x] get byteOffset
- [x] constructor
- [x] copyWithin
- [x] entries
- [x] every
- [x] fill
- [x] filter
- [x] find
- [x] findIndex
- [x] findLast
- [x] findLastIndex
- [x] forEach
- [x] includes
- [x] indexOf
- [x] join
- [x] keys
- [x] lastIndexOf
- [x] get length
- [x] map
- [x] reduce
- [x] reduceRight
- [x] reverse
- [x] set
- [x] slice
- [x] some
- [x] sort
- [x] subarray
- [x] toLocaleString
- [x] toReversed
- [x] toSorted
- [x] toString
- [x] values
- [x] with
- [x] @@iterator
- [x] get @@toStringTag

### Desired features

- [x] Easily return the inner data as `&[u8]` or `&Vec<u8>`

**Comment by AryaveerSR** [comment]
If I understand correctly, the methods need to call the respective method on `TypedArray` with a reference to the inner value and the arguments, do minor wrapping/unwrapping/conversions etc. and return as a `JsValue` right ?

**Comment by jedel1043** [comment]
> If I understand correctly, the methods need to call the respective method on  `TypedArray` with a reference to the inner value and the arguments, do minor wrapping/unwrapping/conversions etc. and return as a `JsValue` right ?

Yes, but some methods like `toSorted` haven't been implemented for `TypedArray` yet

**Comment by AngeloChecked** [comment]
Hi @jedel1043 and @AryaveerSR 
Are you actively addressing this issue? I am willing to assist if there are no concerns about potential overlap?

Just to get better the issue, are these implementations okay, or am I missing something?
```
/// Calls `TypedArray.prototype.keys()`.
pub fn keys(&self, context: &mut Context<'_>) -> JsResult<JsValue> {
    TypedArray::keys(&self.inner, &[], context)
}

/// Calls `TypedArray.prototype.entries()`.
pub fn entries(&self, context: &mut Context<'_>) -> JsResult<JsValue> {
    TypedArray::entries(&self.inner, &[], context)
}

/// Calls `TypedArray.prototype.findLast()`.
pub fn find_last(&self, predicate: JsFunction, context: &mut Context<'_>) -> JsResult<JsValue> {
    TypedArray::find_last(&self.inner, &[predicate.into()], context)
}
```

**Comment by jedel1043** [comment]
@AngeloChecked Yes, those implementations are good! It would be cool to see `findLast` return `Option<JsValue>`, but I don't think `find` does that, so we can leave that as a future improvement.

**Comment by jedel1043** [comment]
I'm not working on this at the moment; I'm a bit busy implementing other builtins.
@AryaveerSR Are you working on any method from this issue?

**Comment by jedel1043** [comment]
> For example, could you confirm if my intuition is correct:

Yep, that's pretty much on point. Our rule is just to use the type system as strictly as we can when returning values, so if a method only returns an index, we return `usize`, or if the method always returns a string, we return `JsString`.

**Comment by AngeloChecked** [comment]
Hi @jedel1043, sorry for bombarding you with questions: I'm using `boa_examples/jstypedarray.rs` for testing new functions. I found the `entries` function returning an iterator, and I need to make `step` and `value` functions of iterators module public instead of pub(crate). 

```rust
/// boa_engine/src/object/builtins/jstypedarray.rs
/// Calls `TypedArray.prototype.entries()`.
pub fn entries(&self, context: &mut Context<'_>) -> JsResult<JsValue> {
    BuiltinTypedArray::entries(&self.inner.clone().into(), &[], context)
}
```

```rust
/// boa_examples/src/bin/jstypedarray.rs
let mut result = vec![];
let data: Vec<u8> = vec![1, 2, 3];
let array = JsUint8Array::from_iter(data, context)?;
let mut iterator = array.entries(context)?.get_iterator(context, None, None)?;
while !iterator.step(context)? {
    let item = iterator.value(context)?.as_object().unwrap().clone();
    let js_array = JsArray::from_object(item)?;
    let index = js_array.get(0, context)?.as_number().unwrap();
    let value = js_array.get(1, context)?.as_number().unwrap();
    result.push((index, value))
}
assert_eq!(vec![(0f64, 1f64), (1f64, 2f64), (2f64, 4f64)], result);
```


```rust
///boa_engine/src/builtins/iterables/mod.rs
/// [spec]: https://tc39.es/ecma262/#sec-iteratorstep
pub fn step(&mut self, context: &mut Context<'_>) -> JsResult<bool> {
    self.step_with(None, context)
}

/// Gets the current value of the `IteratorRecord`.
pub fn value(&mut self, context: &mut Context<'_>) -> JsResult<JsValue> {
    self.set_done_on_err(|iter| iter.last_result.value(context))
}
```

Commonly can I make api public like this? Or do you have any tips on testing iterators or a better approach?

**Comment by jedel1043** [comment]
@AngeloChecked I think you can skip iterator related methods for now. We haven't designed a good public API for Js iterators, so I'd prefer to wait until we have iterators done instead of rushing and exposing internal APIs that may as well change in the future.

**Comment by SuMayaBee** [comment]
Hello, @jedel1043, I noticed this issue hasn't had any activity in a while. If no one is currently working on it, I would love to pick it up. Could you please assign it to me?"

**Comment by gluzandii** [comment]
Hi @jedel1043 — this PR/issue thread looks like it hasn’t had activity in a while, so I took another look at the JsTypedArray TODOs.

From what I can tell, most of the previously-unimplemented iterator-related items (entries, keys, values, @@iterator, etc.) are implemented now. The only items still missing appear to be:

- toLocaleString
- toReversed
- toSorted
- with
-- Also: toLocaleString seems to exist, but the output doesn’t match what I’m seeing in Node/Bun. It might be formatting/locale handling differences.

I would love to work on this issue.

**Comment by aditsharma14** [comment]
Hi maintainers,

I’d like to work on this issue.

After reviewing the codebase, it looks like the necessary infrastructure is already in place (BufferObject::data() and existing typed array access patterns), so adding as_u8() and as_vec() with proper detached buffer handling and tests seems straightforward.

If this approach sounds good, I can start with a draft PR.
Could you please assign the issue to me?

Thanks!

**Comment by ParthMozarkar** [comment]
Hi! I’m interested in contributing to Boa and noticed that this issue is labeled as a good first issue.

If it’s still available, I’d be happy to work on updating the example GIF in the README to reflect the current project behavior.

Please let me know if that would be helpful or if there are any guidelines I should follow before starting.

**Comment by ghost** [comment]
Hi @jedel1043 ! I've been contributing tests for Set.prototype.intersection (#4977) and would like to help complete the JsTypedArray methods.

I noticed that toReversed, toSorted, and with are still listed as missing. I'd like to start by implementing toReversed. I’ll follow the existing patterns in jstypedarray.rs and ensure it aligns with the ECMAScript spec. May I take this one?

**Comment by Xavrir** [comment]
I'd like to work on the remaining unimplemented methods here (toLocaleString, toReversed, toSorted). I'll start with toReversed and toSorted since they have well-defined spec behavior, and then tackle toLocaleString. Will put up a PR when ready.

**Comment by Xavrir** [comment]
Update: after reading the code, I see toReversed, toSorted, and toLocaleString are already implemented in both the builtins layer and the JsTypedArray wrapper. The checkboxes in the issue are just stale. I'll add test coverage for these three methods and update the tracking issue.
